### PR TITLE
Fixes some PHP errors you would get when analyzing a semi-complex int…

### PR DESCRIPTION
**Pull request by angrybrad**
…erface/inheritence tree.

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | n/a
| Fixed issues  |  #141

### CODE REVIEW

**Comment by cebe** [comment]
which problem are you solving with `if (!empty(...)) {` around a foreach? foreach should just not do anything on empty arrays. If somehow it happens that the content is not an array, that bug should be fixed at the place where it is introduced imo.

**Comment by takobell** [comment]
Yeah, it was `null`, which was causing the `foreach` to fail.  I'll see if I can backtrace it to figure out why it is null.

File: models/Context.php [hunk 1]
```diff
@@ -208,30 +208,32 @@ protected function updateSubInterfaceInheritance($interface)
...
-                $inheritedProperty = $this->inheritPropertyRecursive($p, $class);
-                if (!$inheritedProperty) {
-                    $this->errors[] = [
-                        'line' => $p->startLine,
-                        'file' => $class->sourceFile,
-                        'message' => "Method {$p->name} has no parent to inherit from in {$class->name}.",
-                    ];
-                    continue;
-                }
+        if (!empty($class->properties)) {
```
Comment (takobell) [inline models/Context.php]: This one fixes the null set by default on interfaces here: https://github.com/yiisoft/yii2-apidoc/blob/master/models/InterfaceDoc.php#L44-L45

The empty check either needs to happen here or the property should be set to an empty array.

File: models/Context.php [hunk 2]
```diff
@@ -340,7 +342,7 @@ private function inheritPropertyRecursive($method, $class)
      */
     private function getParents($class)
     {
-        if ($class->parentClass === null || !isset($this->classes[$class->parentClass])) {
+        if ($class instanceof InterfaceDoc || $class->parentClass === null || !isset($this->classes[$class->parentClass])) {
```
Comment (takobell) [inline models/Context.php]: We have `interface ElementQueryInterface extends \yii\db\QueryInterface` and a method on that interface with an `@inheritDoc`:

```
/**
 * @inheritdoc
 *
 * @return ElementInterface|null The resulting element.
 */
public function one($db = null);
```

Eventually `Context->inheritMethodRecursive()` gets around to calling `$this->getParents($class)` where `$class` is of type `InterfaceDoc` (referring to our `ElementQueryInterface` and `$method` is referring to our `one()`.

Maybe the bug is that `inheritMethodRecursive()` shouldn't be calling `$this->getParents()` on interfaces?

File: models/Context.php [hunk 3]
```diff
@@ -353,9 +355,12 @@ private function getParents($class)
     private function getInterfaces($class)
     {
         $interfaces = [];
-        foreach($class->interfaces as $interface) {
-            if (isset($this->interfaces[$interface])) {
-                $interfaces[] = $this->interfaces[$interface];
+
+        if (!empty($class->interfaces)) {
```
Comment (takobell) [inline models/Context.php]: Along the same lines of https://github.com/yiisoft/yii2-apidoc/pull/142/files#r106962261, `Context->inheritMethodRecursive()` gets around to calling `$this->getInterfaces($class)` where $class is of type `InterfaceDoc` (referring to our `ElementQueryInterface` and `$method` is referring to our `one()`.  But `InterfaceDoc` doesn't have a `interfaces` property to loop over.

Which again, makes me think that the proper solution for this and the previous comment might be to for `inheritMethodRecursive()` to not call `$this->getInterfaces()` on `InterfaceDocs` classes?


    $inheritanceCandidates = array_merge(
        $this->getParents($class),
        $this->getInterfaces($class)
    );

File: models/Context.php (hunk 1 above)
Comment (samdark) [reply models/Context.php]: https://github.com/yiisoft/yii2-apidoc/commit/1855afb6ad6810520925e3908994ef96ea06ae99

File: models/Context.php (hunk 2 above)
Comment (samdark) [reply models/Context.php]: @angrybrad sorry that the pull request got lost. I'm getting over issues in this repository and trying to verify everything. I'm unsure how to reproduce the issue to fix it i.e. what code should I try to generate apidoc for in order to see the issue?

File: models/Context.php (hunk 3 above)
Comment (samdark) [reply models/Context.php]: Same as https://github.com/yiisoft/yii2-apidoc/pull/142/files#r315932500

**Comment by samdark** [comment]
@angrybrad Sorry that the issue got lost. I'm going through this extension and tried to reproduce the issue but creating all the interfaces and methods you've mentioned I've got no errors. Do you still have a reproduce-able set of code to verify against?

**Comment by angrybrad** [comment]
@samdark no worries... it was a couple of years ago, but I was running it against our Craft codebase.  I'll go ahead and close this and if I can still reproduce, will try to come up with a reproducible test case.

**Pull request closed without merging**